NIP-41 (#829)
nostr-protocol/nips: 41.md

# NIP-41
## Identity rotation
`draft` `optional` `author:pablof7z`

This NIP introduces a simple way in which a pubkey can migrate to by whitelisting a new pubkey ahead of time.

# Basic concepts
* `kind:1776` whitelists a pubkey.
* `kind:1777` migrates to a previously whitelisted pubkey.
* `kind:1776` and `kind:1777` MUST be opentimestamped [NIP-03](https://github.com/nostr-protocol/nips/blob/master/03.md).

> * When a migration event is published, a 30-day period starts in which a user can publish a competing migration event pointing to an earlier `kind:1776` event. After this period, clients SHOULD automatically update the user's follow list to the new pubkey.

Reviewer: So does this mean events published with the new key aren't considered valid for 30 days? It seems like clients could eagerly respect the migration event and switch immediately to the new key, revoking that trust if a competing migration event gets published.

Author: Perhaps 30-days is too long a time-window, but I think *some* delay is needed, otherwise an attacker publishing a revoking event would get exclusive control of the user's key and make things really difficult.

Adding the ability to have to rollback a migration opens a lot more complexity in my opinion, I think once a migration event is effective it should not be possible to roll it back.

But maybe 30-days is too long and this should be more like 2 weeks. Users can also use watchtowers to timestamp+publish `kind:1777`.

Reviewer 2: I think it should be 3 months. 30 days is not enough time.

Reviewer 2: 30 days is not a long time window, in fact it is a short time window.

Clients that want to get fancy about it can follow both keys during that timeframe and add warnings to the users or something like that, as long as the result after 30 days is the same -- but these tricks don't have to be specified in the NIP.

# NIP-41

> ## Identity recovery

Reviewer 2: ```suggestion
## Key migration
```

# NIP-41

> ## Identity rotation

Reviewer 2: ```suggestion
## Identity recovery
```

`draft` `optional` `author:pablof7z`

> This NIP introduces a simple way in which a pubkey can migrate to by whitelisting a new pubkey ahead of time.

Reviewer 2: ```suggestion
This NIP introduces a simple, best-effort, not guaranteed, way in which a user can notify their followers that they have migrated to a new public key in case the current one is compromised.
```

```json
{
    "pubkey": "pubkey-A",
    "kind": 1776,
    "content": "",
    "tags": [
        [ "p", "pubkey-B" ],
        [ "alt", "pubkey whitelisting event" ]
    ]
}
```

* `.content` SHOULD be ignored. Users might choose to use it to leave a base64 symmetrically-encrypted message of where they left the new key or anything else.
* The event MUST have a single `p` tag listing the whitelisted pubkey.

Multiple `kind:1776` events can exist. All `kind:1776` MUST be opentimestamped following [NIP-3](https://github.com/nostr-protocol/nips/blob/master/03.md).

Relays SHOULD NOT delete `kind:1040` nor `kind:1776` events upon receiving a `kind:5` event.

## Migrating to a pubkey
When the user needs to change keys they sign an event `kind:1777` with the new key and creates a NIP-03 attestation.

```json
{
    "pubkey": "pubkey-B",
    "kind": 1777,
    "content": "",
    "tags": [
        [ "p", "pubkey-A" ],
        [ "e", "<kind:1776-event-id>" ],
        [ "proof", "<kind:1040-event-id>" ],
        [ "alt", "pubkey migration event" ],
        [ "relays", "relay1", "relay2" ],

>         [ "mute" ]

Reviewer 2: do we need this? I feel like this should be a client preference. or clients prompting the user when performing the migration: "also mute the old key?"

it doesn't hurt to have this here, of course, I'm just thinking that every line we can eliminate from the NIP makes is more likely to be read.

Author: hahhaa, yeah, I debated whether to add it or not.

I'll take it out.
